rdata/conversion/to_r.py

```diff
@@ -0,0 +1,339 @@
+        r_type = RObjectType.LANG
+        values = data.elements
+        r_value = (build_r_sym(str(values[0]), encoding=encoding),
+                   build_r_list(values[1:], encoding=encoding,
+                                convert_value=build_r_sym))
+
+        if len(data.attributes) > 0:
+            # The following might work here (untested)
+            # attributes = build_r_list(data.attributes, encoding=encoding)  # noqa: ERA001,E501
+            msg = f"type {r_type} with attributes not implemented"
+            raise NotImplementedError(msg)
+
+    elif isinstance(data, (list, tuple, dict)):
+        r_type = RObjectType.VEC
+        values = list(data.values()) if isinstance(data, dict) else data
+        r_value = [convert_to_r_object(el, encoding=encoding) for el in values]
+
+        if isinstance(data, dict):
+            names = np.array(list(data.keys()), dtype=np.dtype("U"))
+            attributes = build_r_list({"names": names},
+                                      encoding=encoding)
+
+    elif isinstance(data, np.ndarray):
+        if data.dtype.kind in ["O"]:
+            # This is a special case handling only np.array([None])
+            if data.size != 1 or data[0] is not None:
+                msg = "general object array not implemented"
+                raise NotImplementedError(msg)
+            r_type = RObjectType.STR
+            r_value = [build_r_object(RObjectType.CHAR)]
+
+        elif data.dtype.kind in ["S"]:
+            assert data.ndim == 1
+            r_type = RObjectType.STR
+            r_value = [convert_to_r_object(el, encoding=encoding) for el in data]
+
+        elif data.dtype.kind in ["U"]:
+            assert data.ndim == 1
+            data = np.array([s.encode(encoding) for s in data], dtype=np.dtype("S"))
+            return convert_to_r_object(data, encoding=encoding)
+
+        else:
+            r_type = {
+                "b": RObjectType.LGL,
+                "i": RObjectType.INT,
+                "f": RObjectType.REAL,
+                "c": RObjectType.CPLX,
+            }[data.dtype.kind]
+
+            if data.ndim == 0:
+                r_value = data[np.newaxis]
+            elif data.ndim == 1:
+                r_value = data
+            else:
+                # R uses column-major order like Fortran
+                r_value = np.ravel(data, order="F")
+                attributes = build_r_list({"dim": np.array(data.shape)},
+                                          encoding=encoding)
+
+    elif isinstance(data, (bool, int, float, complex)):
+        return convert_to_r_object(np.array(data), encoding=encoding)
+
+    elif isinstance(data, str):
+        r_type = RObjectType.STR
+        r_value = [convert_to_r_object(data.encode(encoding), encoding=encoding)]
+
+    elif isinstance(data, bytes):
+        r_type = RObjectType.CHAR
+        if all(chr(byte) in string.printable for byte in data):
+            gp = CharFlags.ASCII
+        elif encoding == "UTF-8":
```

Comment:
If arbitrary encodings are not supported, maybe `encoding` should be a literal too. This would also prevent users to use a different spelling (e.g. `"utf-8"`). We should also consider which spelling do we want to support (I think in Python `"utf-8"` is more common).

Reply:
That seems good.